simba/optimizer_util.py

```diff
@@ -294,39 +294,53 @@ def get_rotation_soc_util(rot_id, schedule, scenario, soc_data: dict = None):
     return scenario.vehicle_socs[rot.vehicle_id], rot_start_idx, rot_end_idx
 
 
-def get_delta_soc(soc_over_time_curve, soc, time_delta, optimizer: 'StationOptimizer'):
+def get_delta_soc(soc_over_time_curve, soc, duration_min: float):
     """ Return expected SoC lift for a given SoC charging time series, start_soc and time_delta.
 
     :param soc_over_time_curve: Data with columns: time, soc and n rows
     :type soc_over_time_curve: np.array() with shape(n, 2)
     :param soc: start socs
     :type soc: float
-    :param time_delta: time of charging
-    :type time_delta: float
-    :param optimizer: optimizer object
-    :type optimizer: simba.station_optimizer.StationOptimizer
+    :param duration_min: duration of charging in minutes
+    :type duration_min: float
     :return: positive delta of the soc
     :rtype: float
     """
     # units for time_delta and time_curve are assumed to be the same, e.g. minutes
     # first element which is bigger than current soc
-    if time_delta == 0:
+
+    if duration_min == 0:
         return 0
-    soc = max(min(optimizer.args.desired_soc_opps, soc), 0)
+    negative_soc = 0
+    if soc < 0:
+        # keep track of the negative soc to add it later
+        negative_soc = -soc
+        try:
+            gradient = soc_over_time_curve[1, 1] - soc_over_time_curve[0, 1]
+        except IndexError:
+            # If no charge curve exists, i.e. less than 2 elements
+            return 0
+        charge_duration_till_0 = -soc / gradient
+        if charge_duration_till_0 > duration_min:
+            return duration_min * gradient
+        duration_min = duration_min - charge_duration_till_0
+        soc = 0
+
     idx = np.searchsorted(soc_over_time_curve[:, 1], soc, side='left')
```

Comment:
what index is that? When the soc reaches 1? Is there guarding needed if vehicles are not meant to be charged 100%?

Reply:
soc_over_time_curve is data of SoC and time, if the battery would be charged with the charge_curve values. the soc is the initial soc/ argument of the function. e.g. a vehicle soc. The index is the index of the first value which is bigger than this soc. 

guarding is done outside of this function. This has been changed to make the function more versatile / reuseable. The function strictly returns the amount of increased soc for greedy charging for a duration. negative start socs are handled as described earlier